File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/agent/AgentPrepMsgTest.java
package org.hyperledger.indy.sdk.agent;

import org.hyperledger.indy.sdk.IndyIntegrationTestWithSingleWallet;
import org.hyperledger.indy.sdk.InvalidStructureException;
import org.hyperledger.indy.sdk.crypto.Crypto;
import org.hyperledger.indy.sdk.crypto.CryptoJSONParameters;
import org.hyperledger.indy.sdk.signus.Signus;
import org.hyperledger.indy.sdk.signus.SignusJSONParameters;
import org.hyperledger.indy.sdk.signus.SignusResults;
import org.hyperledger.indy.sdk.wallet.Wallet;
import org.hyperledger.indy.sdk.wallet.WalletValueNotFoundException;
import org.json.JSONObject;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

import static org.hamcrest.CoreMatchers.isA;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;


public class AgentPrepMsgTest extends IndyIntegrationTestWithSingleWallet {

	private void checkMessage(String senderVk, byte[] encryptedMsg) throws Exception {
		Wallet.createWallet(POOL, "walletForCheck", TYPE, null, null).get();
		Wallet localWallet = Wallet.openWallet("walletForCheck", null, null).get();

		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(null, MY2_SEED, null, true).toJson();
		SignusResults.CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(localWallet, didJson).get();
		String recipientDid = result.getDid();

		byte[] decryptedMessage = Signus.decryptSealed(localWallet, recipientDid, encryptedMsg).get();
		String decryptedMessageJson = new String(decryptedMessage);
		JSONObject decryptedMsg = new JSONObject(decryptedMessageJson);

		assertTrue(decryptedMsg.getBoolean("auth"));
		assertEquals(senderVk, decryptedMsg.getString("sender"));
		assertNotNull(decryptedMsg.getString("nonce"));
		assertNotNull(decryptedMsg.getString("msg"));

		localWallet.closeWallet().get();
		Wallet.deleteWallet("walletForCheck", null).get();
	}

	@Test
	public void testPrepMsgWorksForCreatedKey() throws Exception {
		String paramJson = new CryptoJSONParameters.CreateKeyJSONParameter(MY1_SEED, null).toJson();
		String senderVk = Crypto.createKey(wallet, paramJson).get();

		byte[] encryptedMsg = Agent.prepMsg(wallet, senderVk, VERKEY_MY2, MESSAGE).get();

		checkMessage(senderVk, encryptedMsg);
	}

	@Test
	public void testPrepMsgWorksForCreatedDid() throws Exception {
		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(null, MY1_SEED, null, false).toJson();
		SignusResults.CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(wallet, didJson).get();
		String senderVk = result.getVerkey();

		byte[] encryptedMsg = Agent.prepMsg(wallet, senderVk, VERKEY_MY2, MESSAGE).get();

		checkMessage(senderVk, encryptedMsg);
	}

	@Test
	public void testPrepMsgWorksForCreatedDidAsCid() throws Exception {
		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(null, MY1_SEED, null, true).toJson();
		SignusResults.CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(wallet, didJson).get();
		String senderVk = result.getVerkey();

		byte[] encryptedMsg = Agent.prepMsg(wallet, senderVk, VERKEY_MY2, MESSAGE).get();

		checkMessage(senderVk, encryptedMsg);
	}

	@Test
	public void testPrepMsgWorksForUnknownSenderVerkey() throws Exception {
		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(WalletValueNotFoundException.class));

		Agent.prepMsg(wallet, VERKEY, VERKEY_MY2, MESSAGE).get();
	}

	@Test
	public void testPrepMsgWorksForInvalidRecipientVk() throws Exception {
		String paramJson = new CryptoJSONParameters.CreateKeyJSONParameter(MY1_SEED, null).toJson();
		String senderVk = Crypto.createKey(wallet, paramJson).get();

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		Agent.prepMsg(wallet, senderVk, INVALID_VERKEY, MESSAGE).get();
	}
}

File: wrappers/dotnet/indy-sdk-dotnet-test/AgentTests/AgentPrepMsgTest.cs
using Hyperledger.Indy.AgentApi;
using Hyperledger.Indy.CryptoApi;
using Hyperledger.Indy.SignusApi;
using Hyperledger.Indy.WalletApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.AgentTests
{
    [TestClass]
    public class AgentPrepMsgTest : IndyIntegrationTestWithSingleWallet
    {
        private async Task CheckMessage(string senderVk, byte[] encryptedMsg)
        {
            await Wallet.CreateWalletAsync(POOL, "walletForCheck", TYPE, null, null);
            var localWallet = await Wallet.OpenWalletAsync("walletForCheck", null, null);

            var didJson = string.Format("{{\"seed\":\"{0}\",\"cid\":true}}", MY2_SEED);
            var result = await Signus.CreateAndStoreMyDidAsync(localWallet, didJson);
            var recipientDid = result.Did;

            var decryptedMessageBytes = await Signus.DecryptSealedAsync(localWallet, recipientDid, encryptedMsg);
            var decryptedMessageJson = Encoding.UTF8.GetString(decryptedMessageBytes);
            var decryptedMsg = JObject.Parse(decryptedMessageJson);


            Assert.IsTrue(decryptedMsg.Value<bool>("auth"));
            Assert.AreEqual(senderVk, decryptedMsg.Value<string>("sender"));
            Assert.IsNotNull(decryptedMsg.Value<string>("nonce"));
            Assert.IsNotNull(decryptedMsg.Value<string>("msg"));
           
            await localWallet.CloseAsync();
            await Wallet.DeleteWalletAsync("walletForCheck", null);
        }

        [TestMethod]
        public async Task TestPrepMsgWorksForCreatedKey()
        {
            var paramJson = string.Format("{{\"seed\":\"{0}\"}}", MY1_SEED);

            var senderVk = await Crypto.CreateKeyAsync(wallet, paramJson);
            var encryptedMsg = await Agent.PrepMsgAsync(wallet, senderVk, VERKEY_MY2, MESSAGE);

            await CheckMessage(senderVk, encryptedMsg);
        }

        [TestMethod]
        public async Task TestPrepMsgWorksForCreatedDid()
        {
            var didJson = string.Format("{{\"seed\":\"{0}\",\"cid\":false}}", MY1_SEED);

            var result = await Signus.CreateAndStoreMyDidAsync(wallet, didJson);
            var senderVk = result.VerKey;

            var encryptedMsg = await Agent.PrepMsgAsync(wallet, senderVk, VERKEY_MY2, MESSAGE);

            await CheckMessage(senderVk, encryptedMsg);
        }


        [TestMethod]
        public async Task TestPrepMsgWorksForCreatedDidAsCid()
        {
            var didJson = string.Format("{{\"seed\":\"{0}\",\"cid\":true}}", MY1_SEED);

            var result = await Signus.CreateAndStoreMyDidAsync(wallet, didJson);
            var senderVk = result.VerKey;

            var encryptedMsg = await Agent.PrepMsgAsync(wallet, senderVk, VERKEY_MY2, MESSAGE);

            await CheckMessage(senderVk, encryptedMsg);
        }

        [TestMethod]
        public async Task TestPrepMsgWorksForUnknownSenderVerkey()
        {
            var ex = await Assert.ThrowsExceptionAsync<WalletValueNotFoundException>(() =>
                Agent.PrepMsgAsync(wallet, VERKEY, VERKEY_MY2, MESSAGE)
            );
        }

        [TestMethod]
        public async Task TestPrepMsgWorksForInvalidRecipientVk()
        {
            var paramJson = string.Format("{{\"seed\":\"{0}\"}}", MY1_SEED);
            var senderVk = await Crypto.CreateKeyAsync(wallet, paramJson);

            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                Agent.PrepMsgAsync(wallet, senderVk, INVALID_VERKEY, MESSAGE)
            );
        }
    }
}
